src/scf/newton.jl

```diff
@@ -0,0 +1,489 @@
+    dV = apply_kernel(basis, δρ; ρ=ρ)
+
+    ## Kδφi = Π(dV * φi)
+    Kδφ = similar(φ)
+    for ik = 1:Nk
+        kpt = basis.kpoints[ik]
+        φk = φ[ik]
+        dVφk = similar(φk)
+
+        for i = 1:size(φk, 2)
+            φk_r = G_to_r(basis, kpt, φk[:,i])
+            dVφk_r = dV[:, :, :, kpt.spin] .* φk_r
+            dVφk[:, i] = r_to_G(basis, kpt, dVφk_r)
+        end
+        Kδφ[ik] = dVφk
+    end
+    # ensure proper projection onto the tangent space
+    proj_tangent(Kδφ, φ)
+end
+
+#
+# Callback and convergence
+#
+
+"""
+Flag convergence as soon as total energy change drops below tolerance
+"""
+function NewtonConvergenceEnergy(tolerance)
+    energy_total = NaN
+
+    function is_converged(info)
+        info.energies === nothing && return false # first iteration
+
+        etot_old = energy_total
+        energy_total = info.energies.total
+        abs(energy_total - etot_old) < tolerance
+    end
+    return is_converged
+end
+:w
+
+"""
+Flag convergence as soon as density change drops below tolerance
+"""
+function NewtonConvergenceDensity(tolerance)
+    info -> (norm(info.ρ_next - info.ρ) * sqrt(info.basis.dvol) < tolerance)
+end
+
+"""
+Default callback function for `newton`, which prints a convergence table
+"""
+function NewtonDefaultCallback()
+    prev_energies = nothing
+    function callback(info)
+        !mpi_master() && return info  # Printing only on master
+        if info.stage == :finalize
+            info.converged || @warn "Newton not converged."
+            return info
+        end
+        collinear = info.basis.model.spin_polarization == :collinear
+
+        if info.n_iter == 1
+            E_label = haskey(info.energies, "Entropy") ? "Free energy" : "Energy"
+            magn    = collinear ? ("   Magnet", "   ------") : ("", "")
+            @printf "n     %-12s      Eₙ-Eₙ₋₁     ρ_next-ρ%s\n" E_label magn[1]
+            @printf "---   ---------------   ---------   --------%s\n" magn[2]
+        end
+        E    = isnothing(info.energies) ? Inf : info.energies.total
+        Δρ   = norm(info.ρ_next - info.ρ) * sqrt(info.basis.dvol)
+        if size(info.ρ_next, 4) == 1
+            magn = NaN
+        else
+            magn = sum(spin_density(info.ρ_next)) * info.basis.dvol
+        end
+
+        Estr   = (@sprintf "%+15.12f" round(E, sigdigits=13))[1:15]
+        prev_E = prev_energies === nothing ? Inf : prev_energies.total
+        ΔE     = prev_E == Inf ? "      NaN" : @sprintf "% 3.2e" E - prev_E
+        Mstr = collinear ? "   $((@sprintf "%6.3f" round(magn, sigdigits=4))[1:6])" : ""
+        @printf "% 3d   %s   %s   %2.2e%s\n" info.n_iter Estr ΔE Δρ Mstr
+        prev_energies = info.energies
+
+        flush(stdout)
+        info
+    end
+end
+
+#
+# Newton algorithm
+#
+
+"""
+    newton_step(basis::PlaneWaveBasis, φ, φproj, res, occ;
+                tol_krylov=1e-12, krylov_verbosity=1)
+
+Perform a newton step : we take as given a planewave set φ and we return the
```

Comment:
so just phiproj = phi? What's the point of allowing it to be different?

Reply:
It's weird that you allow it to be different - eg rhoproj, Hproj instead of just rho, H

Reply:
Well, it was when we where wondering about the difference of newton behavior if you project onto current phi or the converged phi. But I agree it can be removed.